src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -441,6 +446,76 @@ try
 }
 CATCH_RETURN()
 
+// Routine Description:
+// - Returns whether the user set or updated any of the font features to be applied
+bool DxFontRenderData::DidUserSetFeatures() const noexcept
+{
+    return _didUserSetFeatures;
+}
+
+// Routine Description:
+// - Clears our internal feature map and populates it with the list of standard features
+void DxFontRenderData::InitializeFeatureMap()
+{
+    _featureMap.clear();
+    _featureMap = {
+        { L"rlig", 1 },
+        { L"rclt", 1 },
+        { L"locl", 1 },
+        { L"ccmp", 1 },
+        { L"calt", 1 },
+        { L"liga", 1 },
+        { L"clig", 1 },
+        { L"kern", 1 },
+        { L"mark", 1 },
+        { L"mkmk", 1 },
+        { L"dist", 1 }
+    };
+}
+
+// Routine Description:
+// - Updates our internal map of font features with the given features
+// Arguments:
+// - features - the features to update our map with
+void DxFontRenderData::SetFeatures(std::unordered_map<std::wstring_view, uint32_t> features)
+{
+    // Populate our feature map with the standard list
+    InitializeFeatureMap();
+
+    // Update our feature map with the provided features
+    if (!features.empty())
+    {
+        for (const auto [tag, param] : features)
+        {
+            _featureMap[tag] = param;
+        }
+        _didUserSetFeatures = true;
+    }
+
+    // Convert the data to DWRITE_FONT_FEATURE and store it in a vector for CustomTextLayout
+    _featureVector.clear();
+    for (const auto [tag, param] : _featureMap)
+    {
+        const auto dwriteTag = DWRITE_MAKE_FONT_FEATURE_TAG(gsl::at(tag, 0), gsl::at(tag, 1), gsl::at(tag, 2), gsl::at(tag, 3));
+        _featureVector.push_back(DWRITE_FONT_FEATURE{ dwriteTag, param });
+    }
+}
+
+// Routine Description:
+// - Updates our internal map of font axes with the given axes
+// Arguments:
+// - axes - the axes to update our map with
+void DxFontRenderData::SetAxes(std::unordered_map<std::wstring_view, int32_t> axes)
+{
+    _axesMap.clear();
+
+    // Update our axis map with the provided axes
+    for (const auto [axis, value] : axes)
+    {
+        _axesMap[axis] = value;
+    }

```

Comment:
If you map it directly to a `_axesVector` here you can skip the work in `_BuildTextFormat`.

Reply:
Good call! Changed